### PR TITLE
Fixes #2855 - app creation failing, if path contains same name multiple times

### LINKED ISSUE #2855: Create app failed when there're multiple same word in the app id
The error log like:

INFO Upgrade id:/someword/dev2/someword/test/111 version:2015-12-17T09:59:37.729Z with force:true (mesosphere.marathon.state.GroupManager:136)
[2015-12-17 17:59:42,867] WARN  (mesosphere.marathon.api.MarathonExceptionMapper$$EnhancerByGuice$$e73f5c3c:30)
java.util.NoSuchElementException: None.get
    at scala.None$.get(Option.scala:347)
    at scala.None$.get(Option.scala:345)
    at mesosphere.marathon.api.v2.AppsResource.replace(AppsResource.scala:166)
    at mesosphere.marathon.DebugModule$MetricsBehavior$$anonfun$invoke$1.apply(DebugConf.scala:50)
    at mesosphere.marathon.metrics.Metrics.timed(Metrics.scala:24)
    at mesosphere.marathon.DebugModule$MetricsBehavior.invoke(DebugConf.scala:49)
    at sun.reflect.GeneratedMethodAccessor264.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60)
    at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:205)
    at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75)
    at com.codahale.metrics.jersey.InstrumentedResourceMethodDispatchProvider$TimedRequestDispatcher.dispatch(InstrumentedResourceMethodDispatchProvider.java:30)
    at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302)
    at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147)
    at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108)
    at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147)
    at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84)
    at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542)
    at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473)
    at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419)
    at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409)
    at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409)
    at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:540)
    at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:715)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:848)
    at com.google.inject.servlet.ServletDefinition.doService(ServletDefinition.java:263)
    at com.google.inject.servlet.ServletDefinition.service(ServletDefinition.java:178)
    at com.google.inject.servlet.ManagedServletPipeline.service(ManagedServletPipeline.java:91)
    at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:62)
    at mesosphere.marathon.api.CacheDisablingFilter.doFilter(CacheDisablingFilter.scala:18)
    at mesosphere.marathon.DebugModule$MetricsBehavior$$anonfun$invoke$1.apply(DebugConf.scala:50)
    at mesosphere.marathon.metrics.Metrics.timed(Metrics.scala:24)
    at mesosphere.marathon.DebugModule$MetricsBehavior.invoke(DebugConf.scala:49)
    at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
    at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
    at mesosphere.marathon.api.CORSFilter.doFilter(CORSFilter.scala:46)
    at mesosphere.marathon.DebugModule$MetricsBehavior$$anonfun$invoke$1.apply(DebugConf.scala:50)
    at mesosphere.marathon.metrics.Metrics.timed(Metrics.scala:24)
    at mesosphere.marathon.DebugModule$MetricsBehavior.invoke(DebugConf.scala:49)
    at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
    at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
    at mesosphere.marathon.api.SetRequestDefaultsFilter.doFilter(SetRequestDefaultsFilter.scala:36)
    at mesosphere.marathon.DebugModule$MetricsBehavior$$anonfun$invoke$1.apply(DebugConf.scala:50)
    at mesosphere.marathon.metrics.Metrics.timed(Metrics.scala:24)
    at mesosphere.marathon.DebugModule$MetricsBehavior.invoke(DebugConf.scala:49)
    at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
    at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
    at mesosphere.marathon.api.LimitConcurrentRequestsFilter.pass(LimitConcurrentRequestsFilter.scala:32)
    at mesosphere.marathon.api.LimitConcurrentRequestsFilter$$anonfun$3$$anonfun$apply$2.apply(LimitConcurrentRequestsFilter.scala:15)
    at mesosphere.marathon.api.LimitConcurrentRequestsFilter$$anonfun$3$$anonfun$apply$2.apply(LimitConcurrentRequestsFilter.scala:15)
    at mesosphere.marathon.api.LimitConcurrentRequestsFilter.doFilter(LimitConcurrentRequestsFilter.scala:36)
    at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
    at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
    at mesosphere.marathon.api.LeaderProxyFilter.doFilter(LeaderProxyFilter.scala:127)
    at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
    at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
    at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:118)
    at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:113)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:501)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:429)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
    at com.codahale.metrics.jetty8.InstrumentedHandler.handle(InstrumentedHandler.java:192)
    at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:154)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
    at org.eclipse.jetty.server.Server.handle(Server.java:370)
    at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:494)
    at org.eclipse.jetty.server.AbstractHttpConnection.content(AbstractHttpConnection.java:982)
    at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.content(AbstractHttpConnection.java:1043)
    at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:865)
    at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:240)
    at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
    at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:696)
    at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:53)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
    at java.lang.Thread.run(Thread.java:744)

### PR DESCRIPTION
**Pull request by alexanderweber**
Check for rest of a path seems not necessary at this point. The recursive function `makeGroup` receives a `gid` which is already rest of the path. Fixes #2855.
@aquamatthias Could you have a look if I am right with my conclusion? I ran all test locally and they are ok.

### CODE REVIEW

**Comment by aquamatthias** [comment]
@alexanderweber looks good. can you please add a specific test case for this scenario?

**Comment by alexanderweber** [comment]
Good point, I've added a small test containing the issue.

File: src/test/scala/mesosphere/marathon/state/GroupTest.scala [hunk 1]
```diff
@@ -376,5 +376,26 @@ class GroupTest extends FunSpec with GivenWhenThen with Matchers {
       Then("the cycle is detected")
       current.hasNonCyclicDependencies should equal(false)
     }
+
+    it("can contain a path which has the same name multiple times in it") {
```
Comment (gkleiman) [inline src/test/scala/mesosphere/marathon/state/GroupTest.scala]: In order to avoid regressions and to make sure that your change fixes the issue, you should add a test that fails without your changes.

I added this test (without the other changes in the commit) to the current master and it passes.

File: src/main/scala/mesosphere/marathon/state/Group.scala [hunk 2]
```diff
@@ -106,12 +106,11 @@ case class Group(
...
-    val restPath = gid.restOf(id)
-    if (gid.isEmpty || restPath.isEmpty) this //group already exists
+    if (gid.isEmpty) this //group already exists
     else {
-      val (change, remaining) = groups.partition(_.id.restOf(id).root == restPath.root)
-      val toUpdate = change.headOption.getOrElse(Group.empty.copy(id = id.append(restPath.rootPath)))
-      val nestedUpdate = if (restPath.isEmpty) toUpdate else toUpdate.makeGroup(restPath.child)
+      val (change, remaining) = groups.partition(_.id.restOf(id).root == gid.root)
+      val toUpdate = change.headOption.getOrElse(Group.empty.copy(id = id.append(gid.rootPath)))
+      val nestedUpdate = if (gid.isEmpty) toUpdate else toUpdate.makeGroup(gid.child)
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: Once you reach this line, you're sure that `gid` is not empty. See line 109.

**Comment by gkleiman** [comment]
@alexanderweber thanks! I added some comments, please re-assign me once you address them.

**Comment by gkleiman** [comment]
@aquamatthias we should set the milestone for this PR. Probably v0.16.0?

File: src/main/scala/mesosphere/marathon/state/Group.scala (hunk 2 above)
Comment (alexanderweber) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: Thanks for pointing that out, I did miss that one. Your suggested snipped causes the equivalent problem with multiple names though. It is due to the `restOf` function, which compares the head of both PathId, cutting off relevant subgroups, which happen to have the same name. By the recursive nature of `makeGroup`, we loose information about the initial parent path with each recursion step. We need the initial path though, in order to distinguish the depth of the path and find the actual rest of value. Instead I would suggest following:
```scala
this.copy(groups = remaining + toUpdate.makeGroup(id.child))
```
So basically getting rid of the if statement. Through the recursion we will step into returning `this` anyway, so it should be equivalent to still having the if statement.

File: src/test/scala/mesosphere/marathon/state/GroupTest.scala (hunk 1 above)
Comment (alexanderweber) [inline src/test/scala/mesosphere/marathon/state/GroupTest.scala]: Good call, such a test would be useless. I will do a new one using the current master as reference.

**Comment by alexanderweber** [comment]
@gkleiman I wrote a new test case. It is failing with the current master and is successful with the PR branch. Could you have a look at my changes in `Group.makeGroup`? As for the tests being successful, they are my only evidence of being correct.

File: src/test/scala/mesosphere/marathon/state/GroupTest.scala [hunk 3]
```diff
@@ -376,5 +376,26 @@ class GroupTest extends FunSpec with GivenWhenThen with Matchers {
...
+      ))
+
+      When("App is updated")
+      val app = AppDefinition("/test/service/test/app".toPath, cmd = Some("Foobar"))
+      val group = Group(PathId("/"), Set(app))
+      val updatedGroup = group.updateApp(app.id, { a => app }, Timestamp.zero)
+      val ids = updatedGroup.transitiveGroups.map(_.id)
+
+      Then("All non existing subgroups should be created")
+      reference.transitiveGroups.map(_.id).forall(ids.contains) should be(true)
```
Comment (gkleiman) [inline src/test/scala/mesosphere/marathon/state/GroupTest.scala]: You will get better errors in case of failures if you change the assertion to something like:

```scala
ids should equal(reference.transitiveGroups.map(_.id).forall(ids.contains))
````

This will display the diff between the two sets instead of something like "false is not true".

**Comment by gkleiman** [comment]
@alexanderweber merged, thanks! :rocket:

File: src/test/scala/mesosphere/marathon/state/GroupTest.scala (hunk 3 above)
Comment (alexanderweber) [inline src/test/scala/mesosphere/marathon/state/GroupTest.scala]: @gkleiman I agree on that. Wanted to change it as soon as I get to the office, but looks like you beat me to it. Thanks for the review again!
